toolchain/lex/tokenized_buffer.h

```diff
@@ -225,20 +207,36 @@ class TokenizedBuffer : public Printable<TokenizedBuffer> {
  private:
   friend class Lexer;
 
-  // A diagnostic location converter that maps token locations into source
-  // buffer locations.
-  class SourceBufferDiagnosticConverter
-      : public DiagnosticConverter<const char*> {
+  class SourcePointerDiagnosticEmitter : public DiagnosticEmitter<const char*> {
    public:
-    explicit SourceBufferDiagnosticConverter(const TokenizedBuffer* tokens)
-        : tokens_(tokens) {}
+    explicit SourcePointerDiagnosticEmitter(DiagnosticConsumer* consumer,
+                                            Lex::TokenizedBuffer* tokens)
+        : DiagnosticEmitter(consumer), tokens_(tokens) {}
+
+   protected:
+    auto ConvertLoc(const char* loc, ContextFnT /*context_fn*/) const
+        -> ConvertedDiagnosticLoc override {
+      return tokens_->SourcePointerToDiagnosticLoc(loc);
+    }
 
-    // Implements `DiagnosticConverter::ConvertLoc`.
-    auto ConvertLoc(const char* loc, ContextFnT context_fn) const
-        -> ConvertedDiagnosticLoc override;
+   private:
+    Lex::TokenizedBuffer* tokens_;
+  };
+
+  class TokenDiagnosticEmitter : public DiagnosticEmitter<Lex::TokenIndex> {
+   public:
+    explicit TokenDiagnosticEmitter(DiagnosticConsumer* consumer,
+                                    Lex::TokenizedBuffer* tokens)
+        : DiagnosticEmitter(consumer), tokens_(tokens) {}
+
+   protected:
+    auto ConvertLoc(Lex::TokenIndex token, ContextFnT /*context_fn*/) const
+        -> ConvertedDiagnosticLoc override {
+      return tokens_->TokenToDiagnosticLoc(token);
+    }
 
    private:
-    const TokenizedBuffer* tokens_;
+    Lex::TokenizedBuffer* tokens_;
```

Comment:
`const` doesn't work here any more? Not a huge deal, just a bit surprising.

Reply:
Fixed (also the unnecessary namespace)